src/GraphQL/Federation/Resolvers/EntityResolver.cs

```diff
@@ -0,0 +1,213 @@
+                value = ValueConverter.ConvertTo(value, prop.PropertyType);
+            // set the property value
+            prop.SetValue(obj, value);
+        }
+        return obj;
+    }
+
+    /// <summary>
+    /// Deserializes a value based on the graph type and value provided.
+    /// </summary>
+    private static object? Deserialize(string fieldName, IGraphType graphType, Type valueType, object? value)
+    {
+        // unwrap non-null graph types
+        if (graphType is NonNullGraphType nonNullGraphType)
+        {
+            if (value == null)
+                throw new InvalidOperationException($"The non-null field '{fieldName}' has a null value.");
+            graphType = nonNullGraphType.ResolvedType!;
+        }
+
+        // handle null values
+        if (value == null)
+            return null;
+
+        // loop through list graph types and deserialize each element in the list
+        if (graphType is ListGraphType listGraphType)
+        {
+            // cast/convert value to an array (it should already be an array)
+            // note: coercing scalars to an array of a single element is not applicable here
+            var array = (value as IEnumerable
+                ?? throw new InvalidOperationException($"The field '{fieldName}' is a list graph type but the value is not a list"))
+                .ToObjectArray();
+            // get the list converter and element type for the list type
+            var listConverter = ValueConverter.GetListConverter(valueType);
+            var elementType = listConverter.ElementType;
+            // deserialize each element in the array
+            for (int i = 0; i < array.Length; i++)
+            {
+                array[i] = Deserialize(fieldName, listGraphType.ResolvedType!, elementType, array[i]);
+            }
+            // convert the array to the intended list type
+            return listConverter.Convert(array);
+        }
+
+        // handle scalar graph types
+        if (graphType is ScalarGraphType scalarGraphType)
+            return scalarGraphType.ParseValue(value);
+
+        // handle object graph types
+        if (graphType is IObjectGraphType objectGraphType)
+        {
+            if (value is not Dictionary<string, object?> dic)
+                throw new InvalidOperationException($"The field '{fieldName}' is an object graph type but the value is not a dictionary");
+
+            return ToObject(valueType, objectGraphType, dic);
+        }
+
+        // union and interface types are not supported
+        throw new InvalidOperationException($"The field '{fieldName}' is not a scalar or object graph type.");
+    }
+
+    private record RepresentationDataLoader(IResolveFieldContext Context, Representation Representation) : IDataLoaderResult
+    {
+        public Task<object?> GetResultAsync(CancellationToken cancellationToken = default) => Representation.Resolver.ResolveAsync(Context, Representation.Value).AsTask();
```

Comment:
I thought `record`s were for data carrier types. Why not `class`?

Reply:
I wanted to use the shorter syntax for initialization .... but that's available with `class` too.  I'll change it.

Reply:
Done